### PR TITLE
docs: add documentation for source persistence

### PR DESCRIPTION
**Pull request by ruchirK**
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4224)

### CODE REVIEW

File: doc/user/content/cli/_index.md [hunk 1]
```diff
@@ -200,3 +202,61 @@ etc.), and then create a new node with those items.
...
+
+### Persistence
+
+{{< warning >}}
+
+Persistence is a experimental feature at this time. Please don't use it
+with any data you are not comfortable losing. To use persistence you need to have
+both the `--persistence` and `--experimental` flags enabled.
+
+{{< /warning >}}
```
Comment (benesch) [inline doc/user/content/cli/_index.md:214]: Can you adjust this to use the experimental shortcode instead, for consistent wording? There's an example in `sql/create-table.md` that you can follow!

File: doc/user/content/cli/_index.md [hunk 2]
```diff
@@ -200,3 +202,61 @@ etc.), and then create a new node with those items.
...
+
+Materialize lets you persist input data from Kafka sources to files on the
```
Comment (sploiselle) [inline doc/user/content/cli/_index.md:216]: One of my technical writing tricks is to shape things as "[What the user wants to accomplish], [how you do it in our product]." This increases the likelihood that the person reading the docs sees that we can help them achieve some  task. So I might suggest rewording this into something like:

> To avoid re-reading data from Kafka on restart, Materialize lets you persist input data from Kafka sources to files on the Materialize instance's local hard drive.

File: doc/user/content/cli/_index.md [hunk 3]
```diff
@@ -200,3 +202,61 @@ etc.), and then create a new node with those items.
...
+Materialize instance's local hard drive, so that subsequent restarts can read the
+previously ingested data from those local files instead of reading the data from
+Kafka again. Currently, source persistence will not solve many restart time or
+memory usage related problems, as it presently only copies data directly from
+Kafka and stores them in local files.
+
+We recommend using persistence if you are using real-time consistency Kafka sources,
```
Comment (sploiselle) [inline doc/user/content/cli/_index.md:223]: Do we have anything in the docs on "real-time consistency Kafka sources" that you can link to? How will someone know if this applies to them or not?

File: doc/user/content/cli/_index.md [hunk 4]
```diff
@@ -200,3 +202,61 @@ etc.), and then create a new node with those items.
...
+need to relieve load on upstream Kafka brokers, and are comfortable using
+[experimental](#experimental-mode) features. We will continue iterating on
+persistence and making it support a wider set of source types and improving
+performance and reliability.
+
```
Comment (sploiselle) [inline doc/user/content/cli/_index.md:228]: I'd add another subheading here like `#### Details` just to make it a little easier to wade through. If this doc gets any bigger, we might also consider breaking it out into its own page?

File: doc/user/content/cli/_index.md [hunk 5]
```diff
@@ -200,3 +202,61 @@ etc.), and then create a new node with those items.
...
+Materialize will store one copy of all input data for each persistent Kafka source.
```
Comment (sploiselle) [inline doc/user/content/cli/_index.md:229]: throughout change "will `<foo>`" to "`<foo>`s", e.g. "will store" to "stores" or "will buffer" to "buffers"

File: doc/user/content/cli/_index.md [hunk 6]
```diff
@@ -200,3 +202,61 @@ etc.), and then create a new node with those items.
...
+
+Here, each file stores data for ranges of offsets per `partition-id`. Each file
+stores all the data from `start-offset` (inclusive) to `end-offset` (exclusive).
+Materialize will buffer up to `--persistence-max-pending-records` records in memory
+per source, before flushing them all to disk immediately. Setting this flag to a
+higher value helps Materialize achieve higher ingest and disk write throughput,
+however this also increases the average latency before records are persisted.
+Additionally, Materialize flushes input records to disk every 10 minutes.
+
+{{< warning >}}
```
Comment (sploiselle) [inline doc/user/content/cli/_index.md:250]: I'd place the warning higher just to increase the likelihood that someone sees it.

Comment (ruchirK) [reply doc/user/content/cli/_index.md]: done

File: doc/user/content/cli/_index.md (hunk 5 above)
Comment (ruchirK) [reply doc/user/content/cli/_index.md]: done

File: doc/user/content/cli/_index.md (hunk 4 above)
Comment (ruchirK) [reply doc/user/content/cli/_index.md]: done

File: doc/user/content/cli/_index.md (hunk 3 above)
Comment (ruchirK) [reply doc/user/content/cli/_index.md]: removed the real time bit

File: doc/user/content/cli/_index.md (hunk 2 above)
Comment (ruchirK) [reply doc/user/content/cli/_index.md]: done

File: doc/user/layouts/partials/create-source/connector/kafka/with-options.html [hunk 7]
```diff
@@ -1,5 +1,6 @@
 `client_id` | `text` | Use the supplied value as the Kafka client identifier.
 `group_id_prefix` | `text` | Prefix Materialize Kafka users' `group.id` with the provided value. The resulting `group.id` looks like `<group_id_prefix>materialize-X-Y`, where `X` and `Y` are values that allow multiple concurrent Kafka consumers from the same topic.
+`persistence` | `bool` | Persist data from this source to local files. Currently only supports real-time consistency sources.
```
Comment (ruchirK) [inline doc/user/layouts/partials/create-source/connector/kafka/with-options.html:3]: get rid of rt here too

**Comment by ruchirK** [comment]
I've responded to all of the comments

I have two lingering thoughts

1. I didn't add this to the release notes because I don't exactly know if we want to announce this for 0.5 or do something else? (the experimental warning said it was added in 0.4.2 which is roughly when things got merged in?)

2. not sure if the docs make it clear that in order to use persistence you have to

a) set `--experimental` and `--persistence`

AND

b) create a persistent source by doing
```
CREATE SOURCE ... WITH(persistence = true)
```

and that if you only do one of those steps its a no-op?

**Comment by sploiselle** [comment]
re: your second point, Ruchir, let's add an example to the appropriate page that has an example of creating a persistent source; you can add a note there that for this to work, you must pass both flags

File: doc/user/content/cli/_index.md [hunk 8]
```diff
@@ -200,3 +202,58 @@ etc.), and then create a new node with those items.
...
+higher value helps Materialize achieve higher ingest and disk write throughput,
+however this also increases the average latency before records are persisted.
+Additionally, Materialize flushes input records to disk every 10 minutes.
+
+
+On restart, Materialize reads back all of the records that had been previously
+persisted in offset order, and then continues reading from the upstream source
+for data after the last persisted record in each partition.
+
+[cs]: /sql/create-source
```
Comment (sploiselle) [inline doc/user/content/cli/_index.md:259]: you can link this to the anchor tag for the example you addd

**Review by sploiselle — APPROVED** [review]

File: doc/user/content/cli/_index.md (hunk 8 above)
Comment (ruchirK) [reply doc/user/content/cli/_index.md]: if i add an example it has to be one specific source type right (like kafka + avro)

so should i do that and then just add a link here (even though it doesnt cover all source types) or should i add an example for each source type?

Comment (sploiselle) [reply doc/user/content/cli/_index.md]: We should add an example for each source type that supports persistence; if there's more than one then maybe consider adding a parenthetical with the links or something
